### DISCUSSION TITLE
Push through unconverged Newton steps?

### DISCUSSION THREAD

**Discussion by YangZhao16** [thelfer/tfel#250] [Q&A]
Hi,

I was trying to model the fracture process with Mfront, and I encountered the convergence problem. The newton iteration works fine for some steps, and then at a certain step, the residual oscillates and the iteration reaches the maximum iteration numbers and the simulation is killed.

I suppose the issue could result from the ill-conditioning of the system for a few edge cases, and I wonder if there's any way we can resort to such that we can push through the unconverged newton steps, for example, simply skip the current step and continue the calculation, rather than letting the program to be killed.

Looking forward to your reply!

**Comment by thelfer** [comment]
@YangZhao16 I am not sure to understand your question without further details and/or an example to play with.

I assume that the divergence lies in MFront and that you are using an implicit integration scheme. If so, you can do this kind of things

```cxx
@AdditionalConvergenceChecks{
   if((!converged) && (iter==iterMax)){ // could be iter==iterMax-1, I didn't check
    converged = true;
  }
}
```

Beware if you use the `StandardElasticityBrick` to compute the consistent tangent operator because the jacobian matrix used to automatically compute the consistent tangent operator may be totally wrong and you may want to reset it to zero (this would given the elastic stifness as the consistent tangent operator).

**Reply by YangZhao16** [reply]
Hi @thelfer,

I was actually asking about the convergence of the global system. It seems that I need to deal with FeniCS if I want to force the loading step to continue when a Newton step of the global system does not converge.

I actually realized the issue that might result in the divergence of my calculation. I was trying to update the geometry of the fracture once solution of a given loading step is determined. However, I did that in the @InitializeLocalVariables keyword. I reviewed the procedure to solve nonlinear systems with MFront, and realized that it might be wrong because @InitializeLocalVariables will be called for every Newton step rather than loading step. As a result, the geometry of fractures will change with some intermediate solution in the Newton iteration process.  

May I confirm with you if my reasoning above makes sense to you? If so, is there a way to update information such as state variables just at the beginning of the loading step?

**Reply by thelfer** [reply]
@YangZhao16 Sorry, I treat too much questions to be able to remember each individual details. Would you give me more details about your computation. Are you using phase-field approaches in FEniCS, or a local damage model ?

**Reply by YangZhao16** [reply]
Hi, 

I was using assumed enhanced strain method to model fracture. Simply speaking, one thing I want to do is to use the stress state and a failure criterion, to detect and mark failed elements at the beginning (or the end) of each loading step.

I used to implement that with @InitializeLocalVariables, and it seems that this keyword will be executed at the beginning of each Newton step, not each loading step.

**Reply by thelfer** [reply]
@YangZhao16 Indeed, there is no way for `MFront` to get at which Newton step the behaviour is called. If you want to do such fancy trick, you must act on the FEniCS side. Are you using `mgis.fenics` or did you implement your own Newton method ?

**Reply by YangZhao16** [reply]
Hi @thelfer,

I've written my code using mgis.fenics, by learning from the example you provided for Von-mises plasticity:
https://thelfer.github.io/mgis/web/mgis_fenics_small_strain_elastoplasticity.html

I wonder if you have ideas on how to modify the code such that the id of Newton step can be feed to Mfront as an Auxiliary State variable. 

Will it be easier for the case that the Newton method for the system is explicitly stated, as the other version of the Von-mise example?
https://comet-fenics.readthedocs.io/en/latest/demo/2D_plasticity/vonMises_plasticity.py.html#vonmisesplasticity

**Reply by thelfer** [reply]
For a test on phase-field method, I added the `add_before_update_constitutive_law_callback` to the `AbstractNonlinearProblem` class in `mgis.fencis`. It takes a callable object which is called before the behaviour integration. You may do a lot of things with that. If you want a method after the behaviour integration, just open an issue about it on the `mgis` website.

**Reply by YangZhao16** [reply]
Hi @thelfer,

Thank you for the reply. May I know if you could share me with the section of the code involving the method add_before_update_constitutive_law_callback? I really appreciate it because I feel it hard to learn the syntax from scratch.

**Reply by thelfer** [reply]
```python
.....
    d = Function(Vd, name="Damage")
    
    # MFront displacement problem
    material_u = mf.MFrontNonlinearMaterial("./src/libBehaviour.so",
                                            "PhaseFieldDisplacementSpectralSplit",
                                            hypothesis="plane_strain",
                                            material_properties={"YoungModulus": 3e3,
                                                                 "PoissonRatio": 0.3})
    problem_u = mf.MFrontNonlinearProblem(u, material_u, quadrature_degree=0, bcs=bcu)
    problem_u.register_external_state_variable("Damage", d)

....
 def callback():
        solve_d(method)
.....
    problem_u.add_before_update_constitutive_law_callback(callback)
```

This code define a function named `callback` which call a method named `solve_d` which modifies a function `d` which is used as an external state variable for the mechanical problem. The `callback` function is called before the behaviour integration at each Newton step.
